crates/extensions/c8y_mapper_ext/src/operations/device_profile.rs

```diff
@@ -0,0 +1,176 @@
+use tedge_api::entity_store::EntityExternalId;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::OperationType;
+use tedge_api::CommandStatus;
+use tedge_api::Jsonify;
+use tedge_mqtt_ext::MqttMessage;
+
+impl CumulocityConverter {
+    /// Convert c8y_DeviceProfile JSON over MQTT operation to ThinEdge device_profile command.
+    pub fn convert_device_profile_request(
+        &self,
+        device_xid: String,
+        cmd_id: String,
+        device_profile_request: C8yDeviceProfile,
+        profile_name: String,
+    ) -> Result<Vec<MqttMessage>, CumulocityMapperError> {
+        let entity_xid: EntityExternalId = device_xid.into();
+
+        let target = self.entity_store.try_get_by_external_id(&entity_xid)?;
+
+        let channel = Channel::Command {
+            operation: OperationType::DeviceProfile,
+            cmd_id,
+        };
+        let topic = self.mqtt_schema.topic_for(&target.topic_id, &channel);
+
+        let mut request = DeviceProfileCmdPayload {
+            status: CommandStatus::Init,
+            name: profile_name,
+            operations: Vec::new(),
+        };
+
+        request.add_firmware(device_profile_request.firmware.into());
+        request.add_software(device_profile_request.software.try_into()?);
+        for config in device_profile_request.configuration {
+            request.add_config(config.into());
+        }
+
+        // Command messages must be retained
+        Ok(vec![
+            MqttMessage::new(&topic, request.to_json()).with_retain()
+        ])
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tests::skip_init_messages;
+    use crate::tests::spawn_c8y_mapper_actor;
+    use crate::tests::TestHandle;
+
+    use c8y_api::json_c8y_deserializer::C8yDeviceControlTopic;
+    use serde_json::json;
+    use std::time::Duration;
+    use tedge_actors::test_helpers::MessageReceiverExt;
+    use tedge_actors::Sender;
+    use tedge_mqtt_ext::test_helpers::assert_received_includes_json;
+    use tedge_mqtt_ext::MqttMessage;
+    use tedge_test_utils::fs::TempTedgeDir;
+
+    const TEST_TIMEOUT_MS: Duration = Duration::from_millis(3000);
+
+    #[tokio::test]
+    async fn mapper_converts_device_profile_operation_for_main_device() {
+        let ttd = TempTedgeDir::new();
+        let test_handle = spawn_c8y_mapper_actor(&ttd, true).await;
+        let TestHandle { mqtt, .. } = test_handle;
+        let mut mqtt = mqtt.with_timeout(TEST_TIMEOUT_MS);
+
+        skip_init_messages(&mut mqtt).await;
+
+        // Simulate c8y_DeviceProfile operation delivered via JSON over MQTT
+        mqtt.send(MqttMessage::new(
+            &C8yDeviceControlTopic::topic(&"c8y".try_into().unwrap()),
+            json!({
+                "id": "123456",
+                "profileName": "test-profile",
+                "c8y_DeviceProfile": {
+                    "software": [
+                    {
+                        "softwareType": "apt",
```

Comment:
Drop the `softwareType` from tests as it is not there at the moment. Use `::apt` type suffixes in the version value for any `type` related tests.

Reply:
We can write tests for both (with and without softwareType)...as there is nothing preventing people from creating the device profiles via REST API (instead of the UI).

Reply:
Resolved by 98b7ffbccc7b8118f7a45d4ef553f28b823bfaba